### ISSUE TITLE
Utility macros for types

### ISSUE THREAD

**Issue by gomoripeti** [josefs/Gradualizer#98]
This proposal is only about ergonomics and nothing to do with type-checker logic.
It was inspired by [`erl_types`](https://github.com/erlang/otp/blob/master/lib/hipe/cerl/erl_types.erl) but slightly outdated by recent introduction of `type/1`/`type/2` helper functions.

I propose to add helper macros like `?t_any`/`t_list([?t_integer])` etc, see https://github.com/gomoripeti/Gradualizer/commit/2e1b256e179976eadd53e3fe57ffd16c3fa54813 (this is an outdated commit but before the effort to rebase it I wanted to ask the team if it makes sense or I should just dismiss it forever)

pros:
- originally came up with this to avoid hassle with `erl_anno:new(0)` all over the place
- it also helps avoiding typos like `type(tuple)` when I really mean `type(tuple, any)` (equivalent with `?t_tuple`)
- apart from constructing types (same usage as `type/1/2` functions) it can also be used to match types ignoring second anno field (and optionally ignoring fourth field too)
- it gives a complete list for the gradualizer developer of what `type()` can be (and with an IDE also autocomplete can work quite nicely like `?t_<TAB>`)

cons:
- handling the erl_parse type tuples as records is a hack, which is hidden by macros which is another hack

**Comment by josefs** [comment]
I like the idea of simplifying pattern matching using macros. We'd have to be very careful about normalizing the types in order for pattern matching to work correctly with the annotations, but that's a good thing.

I interested to hear what other people think about this.

**Comment by UlfNorell** [comment]
A possibility is to have separate macros for patterns and expressions. Something like
```erlang
-define(t_any, {type, erl_anno:new(0), any, []}).
-define(t_any_p, {type, _, any, []}).
```
Another, perhaps nicer, option is to have a variant which takes the annotation as an argument that you can use in patterns:
```erlang
-define(t_any(Ann), {type, Ann, any, []}).
-define(t_any(), ?t_any(erl_anno:new(0))).
```

**Comment by gomoripeti** [comment]
The good thing about record syntax is that `#type{name = any, params = []}` matches any value in the anno field (so we don't have to care whether the type was normalised and position removed or not). When constructing the record the not mentioned field (anno) gets the default value, again win-win :)
I think it is an advantage that we don't have to have separate macros for matching and construction (so I wouldn't spoil it). `Ann` as extra macro arg otoh can be ambiguous (eg `?t_list(Ann)` vs `?t_list([?t_integer])`) - but if we only allow extra `Ann` arg with the variant with the most args then its fine (eg `?t_list`, `?t_list([?_integer])`, `?t_list([?t_integer], ?t_nil)`,  `?t_list(Ann, [?t_integer], ?t_nil)`

In the rare case when we want to copy the annotation from one type to the other, we could use the tuple or record representation directly (bit inconsistent look) or have a pair of getter/setter functions (or the extra `Ann` arg to the macro, that Ulf proposed).

**Comment by Zalastax** [comment]
Does this relate to #85?

**Comment by gomoripeti** [comment]
@Zalastax no-no, I just meant these macros for internal use

**Comment by zuiderkwast** [comment]
I like it very much!

@josefs wrote
> We'd have to be very careful about normalizing the types in order for pattern matching to work correctly with the annotations

I think it shouldn't matter. The macros should use `_` for matching the annotation.

Analogous to the `type/1,2` functions (for expressions), I suggest these `?type/1,2` macros for patterns:

```Erlang
-define(type(T), {type, _, T, []}).
-define(type(T, Args), {type, _, T, Args}).
```

I don't think we need macros for expressions, since we have the functions for that. Possibly we could name the macros `type_pat` to make it clear that it's a pattern.

**Comment by zuiderkwast** [comment]
Now, the macros for patterns are added in master and we have the `type/1,2` functions for expressions. Perhaps we can close this?

At first I didn't get Péter's macro trick, but it's pretty clever: using a record to allow the same macro to be used in patterns and in expressions. The annotation is ignored in patterns and gets the annotation `erl_anno:new(0)` in expressions. It is pretty clever, but it wouldn't pass the Gradualizer, unless we fix #121.
